### ISSUE TITLE
same network  , input and optimizer, but get different loss after several iterations

### ISSUE THREAD

**Issue by longweiwei** [open-mmlab/mmaction2#247]
hi, i encounted a strange problem, two identical network(to be precise,one network is a deep copy of another network ).droput is equal to 0, sharing the same input.  the configuration of each optimizer is the same. part  of the main code show as below.
```
# ./mmaction3/mmaction/models/recognizers/recognizer3d.py
def clip_grads(params,grad_clip):
    params = list(
        filter(lambda p: p.requires_grad and p.grad is not None, params))
    if len(params) > 0:
        return clip_grad.clip_grad_norm_(params, **grad_clip)


# define of the second network 
class Net(nn.Module):
    def __init__(self,backbon,head):
        super(Net,self).__init__()
        self.backbon = backbon
        self.head = head
    def forward(self,x):
        x = self.backbon(x)
        x = self.head(x)
        return x
model = None
optimizer = None

@RECOGNIZERS.register_module()
class Recognizer3D(BaseRecognizer):
    """3D recognizer model framework."""
    def forward_train(self, imgs, labels):
        global model
        global optimizer
        """Defines the computation performed at every call when training."""
        imgs = imgs.reshape((-1,) + imgs.shape[2:])
        if model is None:
            model = Net(copy.deepcopy(self.backbone),copy.deepcopy(self.cls_head))
            model.cuda()
            model.train()
            optimizer = optim.SGD(
                model.parameters(),
                lr= 0.01,
                momentum= 0.9,
                weight_decay=0.0001
        output = model(imgs)
        gt_labels = labels.squeeze()
        if gt_labels.shape == torch.Size([]):
            gt_labels = gt_labels.unsqueeze(0)
        comp_loss = model.head.loss(output,gt_labels)
        
        losses = dict()
        x = self.extract_feat(imgs)
        if hasattr(self, 'neck'):
            x, loss_aux = self.neck(x, labels.squeeze())
            losses.update(loss_aux)
        cls_score = self.cls_head(x)
        loss_cls = self.cls_head.loss(cls_score, gt_labels)
        losses.update(loss_cls)

        //weight update of second network
        optimizer.zero_grad()
        comp_loss['loss_cls'].backward()
        clip_grads(model.parameters(),grad_clip=dict(max_norm=40, norm_type=2))
        optimizer.step()

        print(comp_loss['loss_cls'],'       ',losses['loss_cls'])
        return losses
 ...
```
the optimizer of the first network as follows:
```
....
# ./mmaction3/mmaction/apis/train.py
# put model on gpus
    if distributed:
        find_unused_parameters = cfg.get('find_unused_parameters', False)
        # Sets the `find_unused_parameters` parameter in
        # torch.nn.parallel.DistributedDataParallel
        model = MMDistributedDataParallel(
            model.cuda(),
            device_ids=[torch.cuda.current_device()],
            broadcast_buffers=False,
            find_unused_parameters=find_unused_parameters)
    else:
        model = MMDataParallel(
            model.cuda(cfg.gpu_ids[0]), device_ids=cfg.gpu_ids)

    # build runner
    # optimizer = build_optimizer(model, cfg.optimizer)
    optimizer = optim.SGD(
        model.parameters(),
        lr=0.01,
        momentum=0.9,
        weight_decay=0.0001)
....
``` 
part the output is as follows:
```
tensor(4.8671, device='cuda:0', grad_fn=<MulBackward0>)         tensor(4.8671, device='cuda:0', grad_fn=<MulBackward0>)
tensor(5.4843, device='cuda:0', grad_fn=<MulBackward0>)         tensor(5.4843, device='cuda:0', grad_fn=<MulBackward0>)
tensor(6.7874, device='cuda:0', grad_fn=<MulBackward0>)         tensor(6.7874, device='cuda:0', grad_fn=<MulBackward0>)
tensor(8.0645, device='cuda:0', grad_fn=<MulBackward0>)         tensor(8.0646, device='cuda:0', grad_fn=<MulBackward0>)
tensor(8.5082, device='cuda:0', grad_fn=<MulBackward0>)         tensor(8.5092, device='cuda:0', grad_fn=<MulBackward0>)
tensor(9.2198, device='cuda:0', grad_fn=<MulBackward0>)         tensor(9.1695, device='cuda:0', grad_fn=<MulBackward0>)
tensor(9.9751, device='cuda:0', grad_fn=<MulBackward0>)         tensor(9.9748, device='cuda:0', grad_fn=<MulBackward0>)
tensor(10.7431, device='cuda:0', grad_fn=<MulBackward0>)         tensor(10.7773, device='cuda:0', grad_fn=<MulBackward0>)
tensor(11.3149, device='cuda:0', grad_fn=<MulBackward0>)         tensor(11.3147, device='cuda:0', grad_fn=<MulBackward0>)
tensor(11.9542, device='cuda:0', grad_fn=<MulBackward0>)         tensor(11.9992, device='cuda:0', grad_fn=<MulBackward0>)
tensor(11.1789, device='cuda:0', grad_fn=<MulBackward0>)         tensor(11.0418, device='cuda:0', grad_fn=<MulBackward0>)
tensor(11.4374, device='cuda:0', grad_fn=<MulBackward0>)         tensor(11.5714, device='cuda:0', grad_fn=<MulBackward0>)
tensor(11.5342, device='cuda:0', grad_fn=<MulBackward0>)         tensor(11.3407, device='cuda:0', grad_fn=<MulBackward0>)
tensor(10.2100, device='cuda:0', grad_fn=<MulBackward0>)         tensor(10.2324, device='cuda:0', grad_fn=<MulBackward0>)
tensor(11.4401, device='cuda:0', grad_fn=<MulBackward0>)         tensor(11.0003, device='cuda:0', grad_fn=<MulBackward0>)
tensor(9.8553, device='cuda:0', grad_fn=<MulBackward0>)         tensor(9.8986, device='cuda:0', grad_fn=<MulBackward0>)
tensor(9.5002, device='cuda:0', grad_fn=<MulBackward0>)         tensor(9.2319, device='cuda:0', grad_fn=<MulBackward0>)
tensor(7.9033, device='cuda:0', grad_fn=<MulBackward0>)         tensor(7.6935, device='cuda:0', grad_fn=<MulBackward0>)
tensor(8.2079, device='cuda:0', grad_fn=<MulBackward0>)         tensor(7.5507, device='cuda:0', grad_fn=<MulBackward0>)
tensor(1.9962, device='cuda:0', grad_fn=<MulBackward0>)         tensor(1.9446, device='cuda:0', grad_fn=<MulBackward0>)
tensor(8.0533, device='cuda:0', grad_fn=<MulBackward0>)         tensor(6.1229, device='cuda:0', grad_fn=<MulBackward0>)
tensor(7.0879, device='cuda:0', grad_fn=<MulBackward0>)         tensor(8.0190, device='cuda:0', grad_fn=<MulBackward0>)
tensor(7.9284, device='cuda:0', grad_fn=<MulBackward0>)         tensor(6.5795, device='cuda:0', grad_fn=<MulBackward0>)
tensor(7.8347, device='cuda:0', grad_fn=<MulBackward0>)         tensor(7.4337, device='cuda:0', grad_fn=<MulBackward0>)
tensor(6.6924, device='cuda:0', grad_fn=<MulBackward0>)         tensor(6.3805, device='cuda:0', grad_fn=<MulBackward0>)
tensor(8.1306, device='cuda:0', grad_fn=<MulBackward0>)         tensor(7.9693, device='cuda:0', grad_fn=<MulBackward0>)
tensor(7.3484, device='cuda:0', grad_fn=<MulBackward0>)         tensor(6.7294, device='cuda:0', grad_fn=<MulBackward0>)
tensor(5.4379, device='cuda:0', grad_fn=<MulBackward0>)         tensor(9.4658, device='cuda:0', grad_fn=<MulBackward0>)
tensor(5.6425, device='cuda:0', grad_fn=<MulBackward0>)         tensor(6.7552, device='cuda:0', grad_fn=<MulBackward0>)
tensor(6.9743, device='cuda:0', grad_fn=<MulBackward0>)         tensor(9.1101, device='cuda:0', grad_fn=<MulBackward0>)
tensor(7.7424, device='cuda:0', grad_fn=<MulBackward0>)         tensor(7.1821, device='cuda:0', grad_fn=<MulBackward0>)
tensor(3.7089, device='cuda:0', grad_fn=<MulBackward0>)         tensor(3.5759, device='cuda:0', grad_fn=<MulBackward0>)
tensor(5.1526, device='cuda:0', grad_fn=<MulBackward0>)         tensor(5.8825, device='cuda:0', grad_fn=<MulBackward0>)
tensor(4.7222, device='cuda:0', grad_fn=<MulBackward0>)         tensor(2.9085, device='cuda:0', grad_fn=<MulBackward0>)
tensor(7.0358, device='cuda:0', grad_fn=<MulBackward0>)         tensor(5.7582, device='cuda:0', grad_fn=<MulBackward0>)
tensor(5.1632, device='cuda:0', grad_fn=<MulBackward0>)         tensor(5.3117, device='cuda:0', grad_fn=<MulBackward0>)
tensor(5.6248, device='cuda:0', grad_fn=<MulBackward0>)         tensor(5.6034, device='cuda:0', grad_fn=<MulBackward0>)
tensor(4.5183, device='cuda:0', grad_fn=<MulBackward0>)         tensor(4.7259, device='cuda:0', grad_fn=<MulBackward0>)
tensor(5.2148, device='cuda:0', grad_fn=<MulBackward0>)         tensor(5.4969, device='cuda:0', grad_fn=<MulBackward0>)
tensor(7.0340, device='cuda:0', grad_fn=<MulBackward0>)         tensor(7.9538, device='cuda:0', grad_fn=<MulBackward0>)
``` 
according to the output ,it can be clearly seen that in the first few iterations, the loss is the same, as the iteration increase, the difference in loss in more obvious. this is no random seed and the dropout is disabled, so why is there such a difference?
thanks for your help!!

**Comment by dreamerlin** [comment]
Maybe you should try `--deterministic` to set deterministic options for CUDNN backend. https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/utils.py#L79-L81

**Comment by longweiwei** [comment]
> Maybe you should try `--deterministic` to set deterministic options for CUDNN backend. https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/utils.py#L79-L81

I tried it, but it did not work

**Comment by innerlee** [comment]
could you fix the seed in the beginning, and run the program twice.
for the first time, only forward&backward the first network, and bypass the processing of the second network, and for the second time, train the second network only.

**Comment by longweiwei** [comment]
@innerlee  I did an experiment based on your description. I fixed the random seed and did not train the second network. In the two experiments, the input of the network is the same, there is no random part, such as frame selection, cropping, etc., but the loss is not the same.

**Comment by innerlee** [comment]
setting A: train net1, no train net2
setting B: no train net1, train net2

repeat setting A two times, do you get the same results?

**Comment by longweiwei** [comment]
@innerlee  yes, I got the same result. But how does this explain the initial problem?

**Comment by innerlee** [comment]
I guess there are some states not being copied, or some randomness leaked in.
Could you make a minimal re-produce code, so that others could debug?

**Comment by longweiwei** [comment]
@innerlee  I agree with your opinion，But I really can’t find related problem。

**Comment by innerlee** [comment]
could you make a repo for reproduce?

**Comment by longweiwei** [comment]
@innerlee    https://github.com/longweiwei/errorReproduce
This is the address of the code to reproduce the error, thank you for your help。

**Comment by innerlee** [comment]
Thanks, could also provide the command line in the repo's readme? Ideally, there is a one-line script to produce the outputs here https://github.com/longweiwei/errorReproduce/tree/main/mmaction2

**Comment by longweiwei** [comment]
@innerlee  hi! I updated part of the code so that it can be run directly on the command line, but I found that the print() function is invalid in the terminal, and I don’t know how to solve it.

**Comment by innerlee** [comment]
hi @dreamerlin , do you have time to take a look?

**Comment by dreamerlin** [comment]
> hi @dreamerlin , do you have time to take a look?

OK

**Comment by dreamerlin** [comment]
@congee524 you can have a look when u have a time

**Comment by congee524** [comment]
If you want to train two networks at the same time, you could use tools/dist_train.sh and set different master port numbers for different tasks.
Because the code changes a lot, the issue may be shelved until I have time.
